### ISSUE TITLE
SC2039: In POSIX sh, HOSTNAME is undefined.

### ISSUE THREAD

**Issue by graywolf-at-work** [koalaman/shellcheck#1851]
#### For bugs
- Rule Id (if any, e.g. SC1000): 

```
SC2039
```

- My shellcheck version (`shellcheck --version` or "online"): 

```
 +:( $ shellcheck --version
ShellCheck - shell script analysis tool
version: 0.7.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```

- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on tt and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
set -eu

# shellcheck disable=SC2039
: "${HOSTNAME:?}"

echo "$HOSTNAME"
```

#### Here's what shellcheck currently says:

```
+:( $ shellcheck x

In x line 7:
echo "$HOSTNAME"
      ^-------^ SC2039: In POSIX sh, HOSTNAME is undefined.

For more information:
  https://www.shellcheck.net/wiki/SC2039 -- In POSIX sh, HOSTNAME is undefined.

```

#### Here's what I wanted or expected to see:

Nothing since the `disable=` should silence the first warning and from there on the HOSTNAME should be considered as present.

Use case: I'm getting the HOSTNAME set by thing running me (kubernetes), not by bash. So it is always present since it comes from outside. I've tried supressing it but as you can see it's not completely working. My current workaround is `hostname="${HOSTNAME:?}"` and using `$hostname` from there on but I feel like it should not be necessary and the same solution like for other variables (`${xx:?}`) should work.

**Comment by Nightfirecat** [comment]
`disable` directives apply only to the block/statement below them. This seems like expected behavior because you've only disabled that check for one of the two usages.

**Comment by graywolf-at-work** [comment]
Well yeah but after the `: "${HOSTNAME:?}"` shellcheck should understand that the variable is defined. At least that is my understanding by comparing

```
$ cat x
#!/bin/sh
set -eu

echo "$Z"
$ shellcheck -o check-unassigned-uppercase x

In x line 4:
echo "$Z"
      ^-- SC2154: Z is referenced but not assigned.

For more information:
  https://www.shellcheck.net/wiki/SC2154 -- Z is referenced but not assigned.
```

and

```
$ cat x
#!/bin/sh
set -eu

: "${Z:?}"

echo "$Z"
$ shellcheck -o check-unassigned-uppercase x
```

Shellcheck learned that the `Z` will be set so it no longer warns about it. I think it should behave the same for `HOSTNAME`. I mean, sure, in POSIX it is not guaranteed to be present but I'm making explicit check to make sure it is. So after that point the warning does not make sense.

**Comment by Nightfirecat** [comment]
I'm not sure why the second is not giving the same error as the first. [Substring expansion](https://wiki-dev.bash-hackers.org/syntax/pe#substring_expansion) (the `"${var:default}"` notation you're using) does not define a value on an unset or empty variable.

**Comment by graywolf** [comment]
It's not substring expansion. See https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_02 , it is `Indicate Error if Null or Unset.`. So after it passes that line, shellcheck knows that the variable must be set (well, set and not null).

**Comment by koalaman** [comment]
It's a shame to ruin the illusion, but ShellCheck's not actually being clever in that example. There's no warning because ShellCheck only warns once for each unique variable name on the first use, and this first used happened to be guarded by the `${..?}`, thus suppressing the warning. This is a bit of a bug-as-a-feature, since it lets you do exactly what you did.

ShellCheck doesn't warn if the name is explicitly assigned though, so you can do a 
`HOSTNAME="$HOSTNAME"` with a suitably snarky comment.

**Comment by graywolf-at-work** [comment]
That's a good enough workaround, thanks :)
